jetty-fcgi/fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/HttpChannelOverFCGI.java

```diff
@@ -127,12 +218,48 @@ protected void dispatch()
 
     public boolean onIdleTimeout(Throwable timeout)
     {
-        boolean handle = getRequest().getHttpInput().onIdleTimeout(timeout);
+        boolean handle = doOnIdleTimeout(timeout);
         if (handle)
             execute(this);
         return !handle;
     }
 
+    private boolean doOnIdleTimeout(Throwable x)
+    {
+        boolean neverDispatched = getState().isIdle();
+        boolean waitingForContent;
+        synchronized (_contentQueue)
+        {
+            waitingForContent = _contentQueue.isEmpty() || _contentQueue.peek().remaining() == 0;
+        }
+        HttpInput.Content specialContent;
+        synchronized (_contentQueue)
+        {
+            specialContent = _specialContent;
+        }
+        if ((waitingForContent || neverDispatched) && specialContent == null)
+        {
+            x.addSuppressed(new Throwable("HttpInput idle timeout"));
+            synchronized (_contentQueue)
+            {
+                _specialContent = new HttpInput.ErrorContent(x);
+            }
+            return getRequest().getHttpInput().wakeup();
+        }
+        return false;
+    }
+
+    @Override
+    public void recycle()
+    {
+        synchronized (_contentQueue)
+        {
+            _contentQueue.clear();
```

Comment:
Actually, the queue should always be empty. H2 already checks during `recycle` that no content was left unconsumed and throws an `AssertionError` otherwise.

I'm going to replicate this logic here and in HTTP 1.